adapters/repos/db/vector/sharded_locks_test.go

```diff
@@ -108,3 +111,309 @@ func TestShardedLocks_MixedLocks(t *testing.T) {
+		}()
+
+		m.LockAll()
+
+		select {
+		case <-ch:
+		default:
+			require.Fail(t, "should be unlocked")
+		}
+
+		m.UnlockAll()
+	})
+
+	t.Run("LockAll blocks Lock", func(t *testing.T) {
+		t.Parallel()
+		m := NewShardedLocks(5)
+
+		m.LockAll()
+
+		ch := make(chan struct{})
+		go func() {
+			time.Sleep(500 * time.Millisecond)
+			m.UnlockAll()
+
+			close(ch)
+		}()
+
+		m.Lock(1)
+
+		select {
+		case <-ch:
+		default:
+			require.Fail(t, "should be unlocked")
+		}
+
+		m.Unlock(1)
+	})
+
+	t.Run("LockAll blocks RLock", func(t *testing.T) {
+		t.Parallel()
+		m := NewShardedLocks(5)
+
+		m.LockAll()
+
+		ch := make(chan struct{})
+		go func() {
+			time.Sleep(500 * time.Millisecond)
+			m.UnlockAll()
+
+			close(ch)
+		}()
+
+		m.RLock(1)
+
+		select {
+		case <-ch:
+		default:
+			require.Fail(t, "should be unlocked")
+		}
+
+		m.RUnlock(1)
+	})
+
+	t.Run("LockAll blocks LockAll", func(t *testing.T) {
+		t.Parallel()
+		m := NewShardedLocks(5)
+
+		m.LockAll()
+
+		ch := make(chan struct{})
+		go func() {
+			time.Sleep(500 * time.Millisecond)
+			m.UnlockAll()
+
+			close(ch)
+		}()
+
+		m.LockAll()
+
+		select {
+		case <-ch:
+		default:
+			require.Fail(t, "should be unlocked")
+		}
+
+		m.UnlockAll()
+	})
+
+	t.Run("UnlockAll releases all locks", func(t *testing.T) {
+		t.Parallel()
+		m := NewShardedLocks(5)
+
+		m.LockAll()
+		m.UnlockAll()
+
+		m.Lock(1)
+		m.Unlock(1)
+
+		m.RLock(1)
+		m.RUnlock(1)
+	})
+
+	t.Run("RLockAll blocks Lock", func(t *testing.T) {
+		t.Parallel()
+		m := NewShardedLocks(5)
+
+		m.RLockAll()
+
+		ch := make(chan struct{})
+		go func() {
+			time.Sleep(500 * time.Millisecond)
+			m.RUnlockAll()
+
+			close(ch)
+		}()
+
+		m.Lock(1)
+
+		select {
+		case <-ch:
+		default:
+			require.Fail(t, "should be unlocked")
+		}
+
+		m.Unlock(1)
+	})
+
+	t.Run("RLockAll doesn't block/unblock RLock", func(t *testing.T) {
+		t.Parallel()
+		m := NewShardedLocks(5)
+
+		m.RLockAll()
+		m.RLock(1)
+
+		m.RUnlockAll()
+		m.RUnlock(1)
+	})
+
+	t.Run("RLockAll blocks LockAll", func(t *testing.T) {
+		t.Parallel()
+		m := NewShardedLocks(5)
+
+		m.RLockAll()
+
+		ch := make(chan struct{})
+		go func() {
+			time.Sleep(500 * time.Millisecond)
+			m.RUnlockAll()
+
+			close(ch)
+		}()
+
+		m.LockAll()
+
+		select {
+		case <-ch:
+		default:
+			require.Fail(t, "should be unlocked")
+		}
+
+		m.UnlockAll()
+	})
+
+	t.Run("RLockAll doesn't block RLockAll", func(t *testing.T) {
+		t.Parallel()
+		m := NewShardedLocks(5)
+
+		m.RLockAll()
+		m.RLockAll()
+
+		m.RUnlockAll()
+		m.RUnlockAll()
+	})
+
+	t.Run("unlock should wake up next waiting lock", func(t *testing.T) {
+		t.Parallel()
+		m := NewShardedLocks(2)
+
+		m.RLock(1)
+
+		ch1 := make(chan struct{})
+		ch2 := make(chan struct{})
+
+		go func() {
+			defer close(ch1)
+
+			m.Lock(1)
+		}()
+
+		go func() {
+			defer close(ch2)
+
+			time.Sleep(100 * time.Millisecond)
+			m.Lock(1)
+		}()
+
+		time.Sleep(10 * time.Millisecond)
+		m.RUnlock(1)
+
+		<-ch1
+
+		m.Unlock(1)
```

Comment:
missing 2nd `m.Unlock(1)`? just to have same number of locks and unlocks in the test.

Reply:
Good catch